File: Vector/vector.go
package main

import (
	"errors"
	"fmt"
)

type vector struct {
	arr      []int
	size     int
	capacity int
}

func NewVector(capa int) *vector {
	return &vector{
		arr:      make([]int, 0, capa),
		size:     0,
		capacity: capa,
	}

}

func (v *vector) doubleCapacity() {
	newArr := make([]int, v.size, v.capacity)
	copy(v.arr, newArr)
	v.arr = newArr
}
func (v *vector) push_back(num int) {
	if v.size == v.capacity {
		v.doubleCapacity()
	}
	v.arr = append(v.arr, num)
	v.size++

}

func (v *vector) pop() error {
	if v.size == 0 {
		return errors.New("vector is empty")
	}
	v.size--
	return nil
}

func (v *vector) insert(idx, val int) error {
	if idx < 0 || idx > v.capacity {
		return errors.New("index out of range ")
	}

	if v.size == v.capacity {
		v.doubleCapacity()
	}
	for i := v.size - 1; i >= idx; i-- {
		v.arr[i+1] = v.arr[i]
	}

	v.arr[idx] = val
	v.size++
	return nil

}

func (v *vector) print() {
	fmt.Println(v.arr[:v.size])
}

func (v *vector) right_rotation() {
	if v.size == v.capacity {
		v.doubleCapacity()
	}
	last_index := v.arr[v.size-1]
	for i := v.size - 1; i > 0; i-- {
		v.arr[i] = v.arr[i-1]
	}
	v.arr[0] = last_index
}
func (v *vector) left_rotation() {

	if v.size == v.capacity {
		v.doubleCapacity()
	}
	first_index := v.arr[0]
	for i := 0; i < v.size-2; i++ {
		v.arr[i] = v.arr[i+1]

	}
	v.arr[v.size-1] = first_index
}

func (v *vector) right_rotation_with_steps(num int) {

	if v.size == v.capacity {
		v.doubleCapacity()
	}
	num = num % v.size
	for i := 0; i < num; i++ {
		v.right_rotation()
	}

}

func (v *vector) delete_position(idx int) int {
	deleted := v.arr[idx]
	for i := idx; i < v.size-1; i++ {
		v.arr[i] = v.arr[i+1]
	}
	v.size--
	return deleted
}

File: Vector/c++/vector.cpp
#include <iostream>
#include <memory>
using namespace std;

class Vector {
private:
  int capacity;
  int size;
  unique_ptr<int[]> arr;

  void double_capacity() {
    capacity *= 2;
    unique_ptr<int[]> new_arr = make_unique<int[]>(capacity);
    for (int i = 0; i < size; i++) {
      new_arr[i] = arr[i];
    }
    arr = move(new_arr);
  }

public:
  Vector() : capacity(100), size(0), arr(make_unique<int[]>(100)) {}

  Vector(int si)
      : capacity(max(100, si)), size(0), arr(make_unique<int[]>(capacity)) {}

  void push_back(int value) {
    if (size == capacity) {
      double_capacity();
    }
    arr[size] = value;
    size++;
  }

  void right_rotate() {
    if (size > 0) {
      int temp = arr[size - 1];
      for (int i = size - 1; i > 0; i--) {
        arr[i] = arr[i - 1];
      }
      arr[0] = temp;
    }
  }

  void left_rotate() {
    if (size > 0) {
      int temp = arr[0];
      for (int i = 0; i < size - 1; i++) {
        arr[i] = arr[i + 1];
      }
      arr[size - 1] = temp;
    }
  }

  void right_rotate_steps(int steps) {
    while (steps--) {
      right_rotate();
    }
  }

  int pop(int idx) {
    if (idx >= 0 && idx < size) {
      int item = arr[idx];
      for (int i = idx; i < size - 1; i++) {
        arr[i] = arr[i + 1];
      }
      size--;
      return item;
    }
    return -1;
  }

  int improved_search(int idx) {
    for (int i = 0; i < size - 1; i++) {
      if (arr[i] == idx) {
        int temp = arr[i];
        arr[i] = arr[i - 1];
        arr[i - 1] = temp;
        return i;
      }
    }
    return -1;
  }

  void print() {
    for (int i = 0; i < size; i++) {
      cout << arr[i] << " ";
    }
    cout << endl;
  }
};
